zebra-network/src/config.rs

```diff
@@ -176,3 +184,86 @@ impl Default for Config {
         }
     }
 }
+
+impl<'de> Deserialize<'de> for Config {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        #[derive(Deserialize)]
+        #[serde(deny_unknown_fields, default)]
+        struct DConfig {
+            listen_addr: String,
+            network: Network,
+            initial_mainnet_peers: HashSet<String>,
+            initial_testnet_peers: HashSet<String>,
+            peerset_initial_target_size: usize,
+            #[serde(alias = "new_peer_interval")]
+            crawl_new_peer_interval: Duration,
+        }
+
+        impl Default for DConfig {
+            fn default() -> Self {
+                let config = Config::default();
+                Self {
+                    listen_addr: config.listen_addr.to_string(),
+                    network: config.network,
+                    initial_mainnet_peers: config.initial_mainnet_peers,
+                    initial_testnet_peers: config.initial_testnet_peers,
+                    peerset_initial_target_size: config.peerset_initial_target_size,
+                    crawl_new_peer_interval: config.crawl_new_peer_interval,
+                }
+            }
+        }
+
+        let config = DConfig::deserialize(deserializer)?;
+        let listen_addr = match config.listen_addr.parse::<SocketAddr>() {
+            Ok(socket) => Ok(socket),
+            Err(_) => match config.listen_addr.parse::<IpAddr>() {
+                Ok(ip) => {
+                    let port = match config.network {
+                        Network::Mainnet => 8233,
+                        Network::Testnet => 18233,
+                    };
+                    Ok(SocketAddr::new(ip, port))
+                }
+                Err(err) => Err(de::Error::custom(err)),
+            },
+        }?;
+
+        Ok(Config {
+            listen_addr,
+            network: config.network,
+            initial_mainnet_peers: config.initial_mainnet_peers,
+            initial_testnet_peers: config.initial_testnet_peers,
+            peerset_initial_target_size: config.peerset_initial_target_size,
+            crawl_new_peer_interval: config.crawl_new_peer_interval,
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::Config;
+
+    #[test]
+    fn parse_config_listen_addr() {
+        let fixtures = vec![
+            (r#"{"listen_addr": "0.0.0.0"}"#, "0.0.0.0:8233"),
+            (r#"{"listen_addr": "0.0.0.0:9999"}"#, "0.0.0.0:9999"),
+            (
+                r#"{"listen_addr": "0.0.0.0", "network": "Testnet"}"#,
+                "0.0.0.0:18233",
+            ),
+            (
+                r#"{"listen_addr": "0.0.0.0:8233", "network": "Testnet"}"#,
+                "0.0.0.0:8233",
+            ),
+        ];
+
+        for (config, value) in fixtures {
+            let config: Config = serde_json::from_str(config).unwrap();
+            assert_eq!(config.listen_addr.to_string(), value);
+        }
+    }
+}
```

Comment:
Can you move these tests to a `config/tests.rs` module?

That way, incremental builds are faster, because the entire source file can be skipped during non-test builds.